intermediate_source/rpc_tutorial.rst

```diff
@@ -0,0 +1,529 @@
+
+In the training loop, it first creates a distributed autograd context, which
+will help the distributed autograd engine to find gradients and involved RPC
+send/recv functions. Then, it kicks off the forward pass as if it is a local
+model, and run the distributed backward pass. For the distributed backward, you
+only need to specify a list of roots, in this case, it is the loss ``Tensor``.
+The distributed autograd engine will traverse the distributed graph
+automatically and write gradients properly. Next, it runs the ``step``
+API on the distributed optimizer, which will reach out to all involved local
+optimizers to update model parameters. Compared to local training, one minor
+difference is that you don't need to run ``zero_grad()`` because each autograd
+context has dedicated space to store gradients, and as we create a context
+per iteration, those gradients from different iterations will not accumulate to
+the same set of ``Tensors``.
+
+.. code:: python
+
+    def run_trainer():
+        batch = 5
+        ntoken = 10
+        ninp = 2
+
+        nhid = 3
+        nindices = 3
+        nlayers = 4
+        hidden = (
+            torch.randn(nlayers, nindices, nhid),
+            torch.randn(nlayers, nindices, nhid)
+        )
+
+        model = rnn.RNNModel('ps', ntoken, ninp, nhid, nlayers)
+
+        # setup distributed optimizer
+        opt = DistributedOptimizer(
+            optim.SGD,
+            model.parameter_rrefs(),
+            lr=0.05,
+        )
+
+        # train for 10 iterations
+        for epoch in range(10):
+            # create distributed autograd context
+            with dist_autograd.context():
+                inp = torch.LongTensor(batch, nindices) % ntoken
+                hidden[0].detach_()
+                hidden[1].detach_()
+                output, hidden = model(inp, hidden)
+                # run distributed backward pass
+                dist_autograd.backward([output.sum()])
+                # run distributed optimizer
+                opt.step()
+                # not necessary to zero grads as each iteration creates a different
+                # distributed autograd context which hosts different grads
+                print("Training epoch {}".format(epoch))
+
+
+Finally, let's add some glue code to launch the parameter server and the trainer
+processes.
+
+
+.. code:: python
+
+    def run_ps():
+        pass
+
+    def run_worker(name, rank, func, world_size):
+        os.environ['MASTER_ADDR'] = 'localhost'
+        os.environ['MASTER_PORT'] = '29500'
+        rpc.init_rpc(name, rank=rank, world_size=world_size)
+
+        func()
+
+        # block until all rpcs finish
+        rpc.shutdown()
+
+    mp.set_start_method('spawn')
+    ps = mp.Process(target=run_worker, args=("ps", 0, run_ps, 2))
+    ps.start()
+
+    trainer = mp.Process(target=run_worker, args=("trainer", 1, run_trainer, 2))
+    trainer.start()
+    ps.join()
+    trainer.join()
```

Comment:
Would it be useful to show some output of the training process?

Reply:
Yes, let me add some for the `CartPole` RL example at least. For RNN as I am using random data as input, any suggestions on what we could show here?